File: io-hotmoka-service/src/main/java/io/hotmoka/service/models/values/TransactionReferenceModel.java
package io.hotmoka.service.models.values;

import io.hotmoka.beans.InternalFailureException;
import io.hotmoka.beans.references.LocalTransactionReference;
import io.hotmoka.beans.references.TransactionReference;

/**
 * The model of a transaction reference.
 */
public class TransactionReferenceModel {

	/**
	 * The type of transaction.
	 */
	public String type;

	/**
	 * Used at least for local transactions.
	 */
	public String hash;

    /**
     * Builds the model of a transaction reference.
     * 
     * @param reference the transaction reference to copy
     */
    public TransactionReferenceModel(TransactionReference reference) {
    	if (reference instanceof LocalTransactionReference) {
    		this.type = "local";
    		this.hash = reference.getHash();
    	}
    	else
    		throw new InternalFailureException("unexpected transaction reference of type " + reference.getClass().getName());
    }

    public TransactionReferenceModel() {}

    /**
     * Yields the transaction reference having this model.
     * 
     * @return the transaction reference
     */
    public TransactionReference toBean() {
    	if (type == null)
    		throw new InternalFailureException("unexpected null transaction reference type");

    	switch (type) {
    	case "local": return new LocalTransactionReference(hash);
    	default:
    		throw new InternalFailureException("unexpected transaction reference type " + type);
    	}
    }
}

File: io-hotmoka-network-thin-client/io-hotmoka-network-thin-client-js/src/models/values/TransactionReferenceModel.ts
/**
 * The model of a transaction reference.
 */
import {MarshallingContext} from "../../internal/marshalling/MarshallingContext";
import {HotmokaException} from "../../internal/exception/HotmokaException";

export class TransactionReferenceModel {
    /**
     * The type of transaction.
     */
    type: string
    /**
     * Used at least for local transactions.
     */
    hash: string

    constructor(type: string, hash: string) {

        if (!hash) {
            throw new HotmokaException("hash cannot be null")
        }

        if (!type || type !== 'local') {
            throw new HotmokaException("type must be local")
        }

        this.type = type
        this.hash = hash
    }

    public static into(context: MarshallingContext, transactionReference: TransactionReferenceModel): void {
        context.writeTransactionReference(transactionReference)
    }

    /**
     * Marshals an array of transaction references into a given stream.
     * @param transactionReferences the array of transaction references
     * @param context the context holding the stream
     */
    public static intoArray(transactionReferences: Array<TransactionReferenceModel>, context: MarshallingContext): void {
        context.writeCompactInt(transactionReferences.length)
        transactionReferences.forEach(transactionReference => TransactionReferenceModel.into(context, transactionReference))
    }
}